RFC: Generic Filters for Subscriptions and FETCH (#1164)
moq-wg/moq-transport: draft-ietf-moq-transport.md

Objects have expired from cache, their state becomes unknown, and a relay that
handles a downstream request that includes those Objects re-requests them.

#### OBJECT FILTER Parameter

The OBJECT_FILTER parameter(Parameter Type 0x05) MAY appear in SUBSCRIBE,

> SUBSCRIBE_UPDATE, PUBLISH_OK, TRACK_STATUS or FETCH message.  It is a structure

Reviewer: how about adding this to SubscribeNamespace as well ?

Reviewer: adding it to namespace will enable object filtering across the tracks, where the base case is a single subscription . I was suggesting it would be nice to think the filter more generically

> indicating Objects or ranges of Objects that the Publisher will deliver.

Author: Yes, you composite them.

~~~
Object Filter {
  Type (0x5),
  Length (i),

>   Operand and Flag (i),

Reviewer: Where Operator can be one !-, =, >, < . I think this will cover a good set of use-cases.

>   Operand and Flag (i),

Author: I realized < X is also trivially implemented as a range from [0,X]

> ~~~
> Object Filter {
>   Type (0x5),
>   Length (i),
>   Operand and Flag (i),
>   Values (i) ...
> }
> ~~~
> 
> Operand and Flag contains both an enumeration indicating the Operand, with the
> least significant bit indicating if the filter is negated.
> 
> ~~~
> Negated = Operand and Flag & 0x1
> Operand = Operand and Flag >> 1
> ~~~
> 

Author: ```suggestion
~~~
Object Filter {
  Type (0x5),
  Length (i),
  Operand (7),
  Negation Flag (1),
  [Extension ID (i),
  Values (i) ...
}
~~~

```

Operand | Meaning
0x0 | No Filter
0x1 | Group ID
0x2 | Subgroup ID
0x3 | Object ID
0x4 | Extension ID
0x5 | Publisher Priority

> 

Author: ```suggestion

When Negation Flag is 1, the result of the filter is negated to determine which Objects pass the filter.
```

> Values is an array of integers which encode the values of interest.  When
> Operand is Extension ID, the first Value indicates the Extension ID to filter.
> This value MUST indicate an Extension with an integer type.  The remainder of
> the array is a sequence of pairs indicating the start and length of the matching
> range.  The Start is encoded as a delta from the previous End, or from 0 for the
> first element.  The length indicates the number of elements including Start to
> match.  A length of 0 indicates no endpoint.  For example [ 10, 3, 1, 1, 20, 0 ]
> would match values 10, 11, 12, 14 and 34-max.  Only the first (excluding
> Extension ID) or last value can be 0.  If an endpoint receives a 0 value
> anywhere else, it MUST treat the track as Malformed (see {{malformed-tracks}}).
> 

Author: ```suggestion
Extension ID is present only when Operated is Extension ID (value 0x4) and
indicates the Extension to filter. This MUST indicate an Extension with an
integer type.  Objects without the Extension do not pass the filter.

Values is an array of integers which encode the values of interest.  The
remainder of the array is a sequence of pairs indicating the start and length of
the matching range.  The Start is encoded as a delta from the previous End, or
from 0 for the first element.  The length indicates the number of elements
including Start to match.  An odd number of elements indicates the final range
has no endpoint.  For example [ 10, 3, 1, 1, 20 ] would match values 10, 11, 12,
14 and 34-max.  Only the first value can be 0.  If an endpoint receives a 0
value anywhere else or length of the Values array exceeds TBD, it MUST reject
the request with `INVALID_FILTER`.
```

All filters can be removed from a Subscription by sending SUBSCRIBE_UPDATE with
Operand No Filter.

This parameter MAY appear more than once, and filters are cumulative.  When

> filters are applied, delivery rules regarding Subgroups and FETCH responses are

Author: A max seems very sensible.

modified.  Within a Subgroup, the Publisher is allowed to send an Object on a
Subgroup stream even when it is not the next Object if the expected Objects

> did not pass the filter.  In a FETCH response, the Subscriber can only infer

Author: I am not saying you should send stuff that doesn't pass the filter. I was thinking the latter.  For example f you have a subgroup with objects 0-9, and a filter that says only even objects, then you can put [ 0, 2, 4, 5, 8 ] in a single stream, even though you know you are omitting objects.

That said, I think maybe that is a really bad idea, and we should just use the normal logic and the publisher has to reset the stream and open a new one wherever it skips.

> Objects do not exist from their absence if they pass all filters.

Author: Yes, I realized this too -- eg if I filter on Object ID or Group ID, this makes sense, but for other filter types, the subscriber would never know that Object X was not sent because it had the wrong extension value or it just didn't exist.

> This parameter MAY appear more than once, and filters are cumulative.  When
> filters are applied, delivery rules regarding Subgroups and FETCH responses are
> modified.  Within a Subgroup, the Publisher is allowed to send an Object on a
> Subgroup stream even when it is not the next Object if the expected Objects
> did not pass the filter.  In a FETCH response, the Subscriber can only infer
> Objects do not exist from their absence if they pass all filters.

Author: ```suggestion
This parameter MAY appear up to TBD2 times in a request, and filters are
cumulative.  If the parameter appears more than TBD2 times, the endpoint MUST
reject the request with `INVALID_FILTER`. When filters are applied, delivery
rules regarding FETCH responses are modified (see {{message-fetch}}).  In a
filtered FETCH response, the Subscriber can only infer Objects do not exist from
their absence if they pass all filters.  This is only possible when filtering on
Group ID and/or Object ID, as the Subscriber lacks other fields needed to
evaluate the filters.

Publishers MUST respect Subgroup delivery rules even for filtered responses.  If
an Object in the middle of a Subgroup is filtered out, the Publisher MUST NOT
place another Object in that Subgroup stream (see {{closing-subgroup-streams}}).
```

Objects have expired from cache, their state becomes unknown, and a relay that
handles a downstream request that includes those Objects re-requests them.

#### OBJECT FILTER Parameter

The OBJECT_FILTER parameter(Parameter Type 0x05) MAY appear in SUBSCRIBE,
SUBSCRIBE_UPDATE, PUBLISH_OK, TRACK_STATUS or FETCH message.  It is a structure
indicating Objects or ranges of Objects that the Publisher will deliver.

~~~
Object Filter {
  Type (0x5),
  Length (i),
  Operand (7),
  Negation Flag (1),
  [Extension ID (i),
  Values (i) ...
}
~~~

Operand | Meaning
0x0 | No Filter
0x1 | Group ID
0x2 | Subgroup ID
0x3 | Object ID
0x4 | Extension ID
0x5 | Publisher Priority

> When Negation Flag is 1, the result of the filter is negated to determine which Objects pass the filter.
> Extension ID is present only when Operated is Extension ID (value 0x4) and

Author: ```suggestion
When Negation Flag is 1, the result of the filter is negated to determine which
Objects pass the filter.

Extension ID is present only when Operated is Extension ID (value 0x4) and
```

Objects have expired from cache, their state becomes unknown, and a relay that
handles a downstream request that includes those Objects re-requests them.

#### OBJECT FILTER Parameter

The OBJECT_FILTER parameter(Parameter Type 0x05) MAY appear in SUBSCRIBE,
SUBSCRIBE_UPDATE, PUBLISH_OK, TRACK_STATUS or FETCH message.  It is a structure
indicating Objects or ranges of Objects that the Publisher will deliver.

~~~
Object Filter {
  Type (0x5),
  Length (i),
  Operand (7),
  Negation Flag (1),
  [Extension ID (i),
  Values (i) ...
}
~~~

Operand | Meaning
0x0 | No Filter
0x1 | Group ID
0x2 | Subgroup ID
0x3 | Object ID
0x4 | Extension ID
0x5 | Publisher Priority

When Negation Flag is 1, the result of the filter is negated to determine which
Objects pass the filter.

> Extension ID is present only when Operated is Extension ID (value 0x4) and

Author: ```suggestion
Extension ID is present only when Operand is Extension ID (value 0x4) and
```

indicates the Extension to filter. This MUST indicate an Extension with an
integer type.  Objects without the Extension do not pass the filter.

Values is an array of integers which encode the values of interest.  The
remainder of the array is a sequence of pairs indicating the start and length of
the matching range.  The Start is encoded as a delta from the previous End, or
from 0 for the first element.  The length indicates the number of elements
including Start to match.  An odd number of elements indicates the final range
has no endpoint.  For example [ 10, 3, 1, 1, 20 ] would match values 10, 11, 12,
14 and 34-max.  Only the first value can be 0.  If an endpoint receives a 0
value anywhere else or length of the Values array exceeds TBD, it MUST reject
the request with `INVALID_FILTER`.

> All filters can be removed from a Subscription by sending SUBSCRIBE_UPDATE with
> Operand No Filter.

Author: ```suggestion

All filters can be removed from a Subscription by sending SUBSCRIBE_UPDATE with
Operand No Filter.
```

Objects have expired from cache, their state becomes unknown, and a relay that
handles a downstream request that includes those Objects re-requests them.

#### OBJECT FILTER Parameter

The OBJECT_FILTER parameter(Parameter Type 0x05) MAY appear in SUBSCRIBE,
SUBSCRIBE_UPDATE, PUBLISH_OK, TRACK_STATUS or FETCH message.  It is a structure
indicating Objects or ranges of Objects that the Publisher will deliver.

~~~
Object Filter {
  Type (0x5),
  Length (i),
  Operand (7),
  Negation Flag (1),

>   [Extension ID (i),

Reviewer 2: ```suggestion
  [Extension ID (i),]

  Values (i) ...
}
~~~

Operand | Meaning
0x0 | No Filter
0x1 | Group ID
0x2 | Subgroup ID
0x3 | Object ID
0x4 | Extension ID
0x5 | Publisher Priority

When Negation Flag is 1, the result of the filter is negated to determine which
Objects pass the filter.

Extension ID is present only when Operand is Extension ID (value 0x4) and
indicates the Extension to filter. This MUST indicate an Extension with an
integer type.  Objects without the Extension do not pass the filter.

Values is an array of integers which encode the values of interest.  The

> remainder of the array is a sequence of pairs indicating the start and length of

Reviewer 2: What does "the remainder" here mean?

Author: I think a previous version had the first element optionally be the extension Id, but it was pulled out

> remainder of the array is a sequence of pairs indicating the start and length of

Author: ```suggestion
array is a sequence of pairs indicating the start and length of
```

> the matching range.  The Start is encoded as a delta from the previous End, or

Reviewer 3: > The remainder of the array is a sequence of pairs indicating the start and length of
> the matching range

Saying that the second value indicates the length of the _matching range_ is contradictory to the examples, as it suggests the length within which samples will match. For example a start of 10 and a length of 3 indicates that the matching range is from 10..13 , whereas it is intended to be 10..12. 

I think explicitly defining the matching range is more intuitive for users than specifying a non-inclusive final value. For example, if I am filtering on zip code data, and I want 94552, i have to set the filter length to point at the imaginary 94553, which is not a valid zip code. 

So suggestion is modify the definition so that the length is an inclusive match. This changes the example given later and also relaxes the requirement around which values can be zero. To specify matches for 10, 11, 12, 14 and 34-max we would code [10, 2, 2, 0, 20]

from 0 for the first element.  The length indicates the number of elements

> including Start to match.  An odd number of elements indicates the final range

Author: It only saves one value though at the expense of some readability?

> including Start to match.  An odd number of elements indicates the final range

Reviewer 2: Yes. Just a suggestion for consideration. In some other standards this is always done.
